share/sysbuild/cmake/modules/sysbuild_extensions.cmake

```diff
@@ -351,6 +368,93 @@ function(ExternalZephyrProject_Cmake)
   import_kconfig(CONFIG_ ${BINARY_DIR}/zephyr/.config TARGET ${ZCMAKE_APPLICATION})
 endfunction()
 
+# Usage:
+#   ExternalZephyrProject_Group(<name>
+#                               [INCLUDE <images>]
+#                               [EXCLUDE <images>]
+#                               [INCLUDE_FROM <groups>]
+#                               [EXCLUDE_FROM <groups>]
+#                               [IMMUTABLE]
+#   )
+#
+# This function defines groups of external Zephyr projects. Images added using
+# `ExternalZephyrProject_Add()` may inherit common properties of the group(s)
+# they are assigned to.
+#
+# This function can either add a new group "<name>", or extend the definition of
+# an existing group with that name. This lets multiple sources influence what a
+# given group should represent in the multiimage build system. The final set of
+# images/properties for every group is evaluated in `sysbuild_groups_resolve()`.
+#
+# <name>                 Name of the group.
+# INCLUDE <images>:      Add a set of images to the group.
+# EXCLUDE <images>:      Remove a set of images from the group.
+# INCLUDE_FROM <groups>: INCLUDE all images belonging to each group in the list.
+# EXCLUDE_FROM <groups>: EXCLUDE all images belonging to each group in the list.
+# IMMUTABLE:             Disallow extending the definition of this group.
```

Comment:
@tejlmand: https://github.com/zephyrproject-rtos/zephyr/pull/57884#discussion_r1200518047

> At current stage I don't see any reasons why sysbuild should support this.
> 
> It will just result in users having to look into the sysbuild CMake code to understand how to manipulate the right properties, so atm it seems just providing false impression that a group is locked.

To me, these properties are really basic. If they're supported, I think people will use them, so I don't see why we should hold off on these conveniences.

The `IMMUTABLE` property is used right now by the "special groups".

Why do you say that users will have to look at code to understand how these properties work? Do you mean that this function is poorly documented?

Reply:
I think he is specifically talking about the IMMUTABLE option in his comment

Reply:
Exactly, it doesn't lock the group.
Any user who wants to manipulate the group can do so simply by accessing the properties directly.

Hence, `IMMUTABLE` is not possible, which again means that it is giving a false impression to basic users.
Users who might believe groups are indeed immutable.

Reply:
> do you know of a way in CMake to make a "write once then read only" variable or property?

you can do so on normal variables, but I still lack to see the exact purpose.
Are we inventing a feature, where users will come back and ask, `Why can I not adjust group FOO ?`

This is not like a running program where you must prevent changes for security reasons, it is the build system which should provide sufficient flexibility, just like we do with hooks and Zephyr modules.

So far i'm not convinced that this is a core part of the group feature, and i'm not sure who we are helping by providing this flag.

Reply:
The purpose of `IMMUTABLE` is mainly to define special groups like `all` and `board_*`, which are auto-generated and whose contents are guaranteed to be consistent and predictable. Such groups are not meant to be extended; they're meant to be used in definitions of other groups.

For example, it would be abuse to exclude images from `all`, or to include images in `board_mps2_an521` which aren't built for `mps2_an521`.

To keep the implementation of these special groups generic, I made `IMMUTABLE` part of the public API. I made no judgement on whether other people *should* use it, but I figured someone might want to.

I think you both have good reasons for wanting to remove it, but this would require changing how the special groups are represented. They would no longer exist as normal groups, but as something that's only a part of `sysbuild_groups_resolve()`.